In place of an earlier phylogeny-based initial clustering, we used MMseqs2 (coverage mode 1, identity threshold 0.4, and coverage 0.7) to produce initial clusters {​10.1038/s41467-018-04964-5}. These parameters were determined experimentally to keep together hvNLR clades in the published manually curated datasets. The clustered sequences were aligned with MAFFT v7.520 {​10.1093/molbev/mst010}, and cluster phylogenies were built using RAxML-NG {​10.1093/bioinformatics/btz305} with model JTT and 100 bootstrap replicates. The cluster refinement step was performed in a fully automated way using R script CL_AutoRefine.R. It refined clusters iteratively using the following steps: i) identify tree branches to be cut based on ecotype overlap (10 ecotypes), maximum allowed branch length (0.3), and a minimum required bootstrap support (90) ii) split the tree in two by cutting the branch with the highest ecotype overlap, in case of multiple candidates, choose the longest branch iii) update ecotype overlap for the resulting sub-trees iv) repeat until no candidate branches remain that can be cut. Finally, R script CL_Get_Entropy_par.R was used to calculate alignment entropy scores for the sub-cluster alignments. It labeled hvNLRs based on alignments that showed 10 or more positions with entropy of 2 bits or higher.
from the 40 wheat genomes where NLRs were reannotated as described above. The pipeline starting from a collection of NLR sequences and ending with alignments and phylogenetic trees of NLR clusters was executed using Snakemake {​10.12688/f1000research.29032.2}. The Snakefile that documents the steps along with all outputs were uploaded to the project GitHub repository.
