## Exome capture sequencing

Exome capture sequencing data was generated by [Krasileva et al., 2017](https://www.pnas.org/doi/10.1073/pnas.1619268114) and available through [PRJNA258539](https://www.ncbi.nlm.nih.gov/sra?linkname=bioproject_sra_all&from_uid=258539). There are 1,479 SRA experiments deposited under this BioProject. 39 of them are additional libraries deposited existing Kronos mutants. We are therefore targeting to analyze 1,440 mutants. 

We processed about 24 samples in a single batch, and this information is available in **Capture_design.txt**. In this file also shown below, the samples that share the identical 'Capture ID' were processed and sequenced in the same batch by Krasileva et al. Those that share the identical 'Remap ID' were analyzed in the same batch to identify EMS-type mutations. 

```
Mutant ID	SRA accession	Capture ID	Remap ID
Kronos0-1	SRX688079	KTC1	1
Kronos166	SRX688097	KTC1	1
Kronos199	SRX688114	KTC1	1
Kronos447	SRX688221	KTC1	1
Kronos499	SRX688241	KTC1	1
Kronos910	SRX688360	KTC1	1
```

General workflow is provided below:

Obtain SRR accessions for the given SRX accessions.
```
#this can be done with a python script
from pysradb import SRAweb

for line in open('Capture_design.txt', 'r'):
  items = line.split('\t')
  if items[-1] != '#N/A' and int(items[-1]) == RemapID:
    srx = items[1]
    srr = db.srx_to_srr([srx])['run_accession'].iloc[0]
    print(f'{srx} {srr}')
```


${srr} obtained from the previous step can be set up as ${accession}. Download the sequencing data from the Sequence Read Archive with sra-toolkit v3.1.1. 
```
sratoolkit.3.1.1-centos_linux64/bin/prefetch ${accession}
sratoolkit.3.1.1-centos_linux64/bin/fasterq-dump -O . -e ${Numthreads} ${accession}
```

Filter the sequencing data with fastp v0.23.4.
```
fastp --in1 ${accession}_1.fastq --in2 ${accession}_2.fastq --out1 ${accession}.1.filtered.fq --out2 ${accession}.2.filtered.fq --thread 16 -q 20
```

Align the reads to the broken Kronos genome with bwa aln v0.7.18-r1243-dirty.
```
# ${x} is either 1 or 2 for R1 and R2 in the paired-end library
bwa aln -t ${Numthreads} ${reference_dir}/Kronos -f ${accession}.${x}.sai ${accession}.${x}.filtered.fq
bwa sampe -N 10 -n 10 -f ${accession}.sam ${reference_dir}/Kronos ${accession}.1.sai ${accession}.2.sai ${accession}.1.filtered.fq ${accession}.2.filtered.fq
```

Sort the alignment file with samtools v1.20 and remove duplicates with picard v3.0.0
```
samtools view -@ ${Numthreads} -h -b ${accession}.sam | samtools sort -@ ${Numthreads} > ${accession}.sorted.bam
samtools index ${accession}.sorted.bam

#occasionally, the bam file may contains records that picard does not like to process.
#ALIDATION_STRINGENCY=LENIENT can be added to skip those alignments
picard MarkDuplicates REMOVE_DUPLICATES=true I=${accession}.sorted.bam O=${accession}.sorted.rmdup.bam M=${accession}.rmdup.txt
```


